### PR TITLE
Handle timeouts

### PR DESCRIPTION
**Pull request by lhoestq**
As noticed in https://github.com/huggingface/datasets/issues/1939, timeouts were not properly handled when loading a dataset.
This caused the connection to hang indefinitely when working in a firewalled environment cc @stas00 

I added a default timeout, and included an option to our offline environment for tests to be able to simulate both connection errors and timeout errors (previously it was simulating connection errors only).

Now networks calls don't hang indefinitely.
The default timeout is set to 10sec (we might reduce it).

### CODE REVIEW

**Comment by stas00** [comment]
I never said the calls were hanging indefinitely, what we need is quite different - in the firewalled env with a network, there should be no network calls or the should fail instantly.

To make this work I suppose on top of this PR we need:
1. `DATASETS_OFFLINE` env var to force set timeout to 0 globally (or to 0.0001 if 0 has a special meaning of no timeout)
2. `DATASETS_OFFLINE` should guard against failing network calls and not fail the program if it has all the data it needs locally.

Bottom line - if the logic wants to check online if the local file matches online dataset name, let it go wild, but it should fail instantly, recover and use the local file - either specified explicitly or cache if there is one. And only if neither was found only then assert.

I hope this makes sense and is doable.

Thank you, @lhoestq

**Comment by lhoestq** [comment]
Yes that was the first step to add DATASETS_OFFLINE :)

With this PR, if a request times out (which couldn't happen before because no time out was set), it falls back on the local files with no error.

As you said, setting the timeout to something like 1e-16 makes the requests fail instantly, which is one step forward. One last thing left is to disable request retries and everything will be instant !

**Comment by stas00** [comment]
Ah, fantastic. Thank you for elucidating that this PR is part of a bigger master plan!

**Review by SBrandeis — APPROVED** [review]

**Comment by lhoestq** [comment]
Merging this one, then I'll open a new PR for the `DATASETS_OFFLINE` env var :)
